File: app/src/main/java/com/recursia/popularmovies/di/modules/RepositoryModule.java
package com.recursia.popularmovies.di.modules;

import com.recursia.popularmovies.data.db.MovieDao;
import com.recursia.popularmovies.data.mappers.EntityToMovieExtraDatabaseModelMapper;
import com.recursia.popularmovies.data.mappers.MovieDatabaseModelToEntityMapper;
import com.recursia.popularmovies.data.mappers.MovieExtraDatabaseModelToEntityMapper;
import com.recursia.popularmovies.data.network.MoviesApi;
import com.recursia.popularmovies.data.mappers.CreditsResponseToCastMapper;
import com.recursia.popularmovies.data.mappers.DiscoverMovieResponseToMovieMapper;
import com.recursia.popularmovies.data.mappers.MovieTrailersResponseToTrailersMapper;
import com.recursia.popularmovies.data.repositories.MoviesRepositoryImpl;
import com.recursia.popularmovies.domain.MoviesRepository;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

@Module(includes = {RetrofitModule.class, RoomModule.class, MapperModule.class})
public class RepositoryModule {

    @Provides
    @Singleton
    MoviesRepository provideMoviesRepository(MovieDao movieDao,
                                             MoviesApi moviesApi,
                                             MovieDatabaseModelToEntityMapper movieDatabaseModelToEntityMapper,
                                             DiscoverMovieResponseToMovieMapper discoverMovieResponseToMovieMapper,
                                             CreditsResponseToCastMapper creditsResponseToCastMapper,
                                             MovieTrailersResponseToTrailersMapper movieTrailersResponseToTrailersMapper,
                                             MovieExtraDatabaseModelToEntityMapper movieExtraDatabaseModelToEntityMapper,
                                             EntityToMovieExtraDatabaseModelMapper entityToMovieExtraDatabaseModelMapper) {
        return new MoviesRepositoryImpl(
                movieDao,
                moviesApi,
                movieDatabaseModelToEntityMapper,
                discoverMovieResponseToMovieMapper,
                creditsResponseToCastMapper,
                movieTrailersResponseToTrailersMapper,
                movieExtraDatabaseModelToEntityMapper,
                entityToMovieExtraDatabaseModelMapper);
    }
}

File: app/src/main/java/com/recursia/popularmovies/di/modules/RepositoryModule.kt
package com.recursia.popularmovies.di.modules

import com.recursia.popularmovies.data.mappers.*
import com.recursia.popularmovies.data.network.MoviesApi
import com.recursia.popularmovies.data.network.TranslateApi
import com.recursia.popularmovies.data.repositories.AccountRepositoryImpl
import com.recursia.popularmovies.data.repositories.MoviesRepositoryImpl
import com.recursia.popularmovies.data.repositories.TranslateRepositoryImpl
import com.recursia.popularmovies.domain.AccountRepository
import com.recursia.popularmovies.domain.MoviesRepository
import com.recursia.popularmovies.domain.TranslateRepository
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module(includes = [
    MoviesRetrofitModule::class,
    TranslateRetrofitModule::class,
    MapperModule::class])
class RepositoryModule {

    @Provides
    @Singleton
    internal fun provideMoviesRepository(
            moviesApi: MoviesApi,
            discoverMovieResponseToMovieMapper: DiscoverMovieResponseToMovieMapper,
            creditsResponseToCastMapper: CreditsResponseToCastMapper,
            movieTrailersResponseToTrailersMapper: MovieTrailersResponseToTrailersMapper,
            reviewsResponseToReviewMapper: ReviewsResponseToReviewMapper,
            movieModelToEntityMapper: MovieModelToEntityMapper
    ): MoviesRepository {
        return MoviesRepositoryImpl(
                moviesApi,
                discoverMovieResponseToMovieMapper,
                creditsResponseToCastMapper,
                movieTrailersResponseToTrailersMapper,
                reviewsResponseToReviewMapper,
                movieModelToEntityMapper
        )
    }


    @Provides
    @Singleton
    internal fun provideAccountRepository(): AccountRepository {
        return AccountRepositoryImpl()
    }

    @Provides
    @Singleton
    internal fun provideTranslateRepository(translateApi: TranslateApi): TranslateRepository {
        return TranslateRepositoryImpl(translateApi)
    }

}
